6.0 Backport to add support for custom taxonomies filtering for Query Loop (from Gutenberg)

hellofromtonya opened the pull request:
Backports changes from Gutenberg to add support for custom taxonomies filtering [Block Library - Query Loop].

- [X] [`lib/compat/wordpress-6.0/blocks.php`](https://github.com/WordPress/gutenberg/blob/trunk/lib/compat/wordpress-6.0/blocks.php)
- [x] [`lib/compat/wordpress-6.0/block-patterns.php`](https://github.com/WordPress/gutenberg/blob/trunk/lib/compat/wordpress-6.0/block-patterns.php)
- [X] [`lib/compat/wordpress-6.0/block-patterns-update.php`](https://github.com/WordPress/gutenberg/blob/trunk/lib/compat/wordpress-6.0/block-patterns-update.php)
- [x] [`lib/compat/wordpress-6.0/class-wp-rest-block-pattern-categories-controller.php`](https://github.com/WordPress/gutenberg/blob/trunk/lib/compat/wordpress-6.0/class-wp-rest-block-pattern-categories-controller.php)
- [x] [`lib/compat/wordpress-6.0/class-wp-rest-block-patterns-controller.php`](https://github.com/WordPress/gutenberg/blob/trunk/lib/compat/wordpress-6.0/class-wp-rest-block-patterns-controller.php)
- [x] [`lib/compat/wordpress-6.0/class-wp-rest-block-patterns-controller.php`](https://github.com/WordPress/gutenberg/blob/trunk/lib/compat/wordpress-6.0/class-wp-rest-pattern-directory-controller.php)
- [X] `phpunit/class-wp-rest-block-pattern-categories-controller-test.php`
- [X] `phpunit/class-wp-rest-block-patterns-controller-test.php`
- [X] `lib/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php` `WP_Theme_JSON:: get_patterns()`

Also being tracked in [this Gutenberg issue](https://github.com/WordPress/gutenberg/issues/39889).

Trac ticket: https://core.trac.wordpress.org/ticket/55505

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

hellofromtonya committed:
Backports from lib/compat/wordpress-6.0/block-patterns-update.php

hellofromtonya committed:
Backports from lib/compat/wordpress-6.0/blocks.php.

peterwilsoncc commented on src/wp-includes/blocks.php:
```diff
@@ -1126,15 +1126,39 @@ function build_query_vars_from_query_block( $block, $page ) {
+					$term_ids = array_filter( $term_ids );
+
+					$query['tax_query'][] = array(
+						'taxonomy'         => $taxonomy,
+						'terms'            => $terms,
```
`$taxonomy` and `$terms` are unescaped here. The taxonomy also needs to be confirmed as publicly queryable.


```php
if ( ! is_taxonomy_viewable( $taxonomy ) { continue } // This runs it through `taxonomy_exists()` too
```

* If term is a slug, then `array_filter( array_map( 'sanitize_title', $terms ) )`
* If this is a typo and intended to be term_ids, then do `array_filter( array_map( 'intval', $terms ) )` here so the value isn't accidently messed up by our future selves by adding an error on line 1154.

ntsekouras replied:
It's only `ids`, there is no support for `slugs`. I'll see about the validations. I'll make some updates to [the GB PR](https://github.com/WordPress/gutenberg/pull/39970) and will ping you as well then to take a look.

ntsekouras replied:
Also noting that the `$terms` usage was a bug and will be replaced with `$term_ids`, which are sanitized.

hellofromtonya replied:
Resolved with https://github.com/WordPress/gutenberg/pull/39970 in https://github.com/WordPress/wordpress-develop/commit/e51fd0c51ea297bbc717cb42b344b956d0f3f8b4.

peterwilsoncc commented on src/wp-includes/blocks.php:
```diff
@@ -1126,15 +1126,39 @@ function build_query_vars_from_query_block( $block, $page ) {
-			$term_ids              = array_filter( $term_ids );
-			$query['category__in'] = $term_ids;
+		// Migrate `categoryIds` and `tagIds` to `tax_query` for backwards compatibility.
+		if ( ! empty( $block->context['query']['categoryIds'] ) || ! empty( $block->context['query']['tagIds'] ) ) {
```
I can't see why this outer if is needed, the same checks are being done inside it.

ntsekouras replied:
Just to  'encapsulate' (let's say) the 'migration' code. Also, both old attributes merge into a single one and without the extra if we would have more checks about new value and merge, etc..

I think there is no harm to leave this as is.

peterwilsoncc commented on src/wp-includes/blocks.php:
```diff
@@ -1126,15 +1126,39 @@ function build_query_vars_from_query_block( $block, $page ) {
-			$term_ids         = array_filter( $term_ids );
-			$query['tag__in'] = $term_ids;
+		if ( ! empty( $block->context['query']['taxQuery'] ) ) {
+			$query['tax_query'] = array();
```
This looks like it will stomp on any categories and tags in the shim above if the query block has both a legacy value and `taxQuery` defined.

ntsekouras replied:
I'm not sure if it will be possible to have the old and new attributes, as there is the client parsing that migrates the block. So in the case of deprecated blocks, we just fill the new attribute from the old ones.

peterwilsoncc replied:
> I'm not sure if it will be possible to have the old and new attributes, as there is the client parsing that migrates the block.

Thanks, I gather that will come in as part of a package.

adamziel replied:
Using an `else if` instead of an `if` would make this explicit for the future readers of this code. I'm thinking:
```php
else if ( ! empty( $block->context['query']['taxQuery'] ) ) {
```
Not a blocker for the backport, though.

peterwilsoncc commented on src/wp-includes/blocks.php:
```diff
@@ -1126,15 +1126,39 @@ function build_query_vars_from_query_block( $block, $page ) {
-			$term_ids         = array_map( 'intval', $block->context['query']['tagIds'] );
-			$term_ids         = array_filter( $term_ids );
-			$query['tag__in'] = $term_ids;
+		if ( ! empty( $block->context['query']['taxQuery'] ) ) {
```
validation: is_array or is_object or whatever

ntsekouras replied:
`$block->context['query']['taxQuery']` is an `object`, but we have similar checks above without checking the type. Do you think we need to update every similar check? I don't think there is any security issue here, is there?

peterwilsoncc replied:
Thanks Nik, I think it's fine. The only risk I can see is that PHP with throw an error if the type is unexpected.

peterwilsoncc reviewed (commented):
Added some notes inline

gziolo commented:
@ntsekouras can you help with backporting this functionality?

gziolo commented:
I also see files listed in the description that @jsnajdr added, so let's keep him in the loop.

ntsekouras commented:
>@ntsekouras can you help with backporting this functionality?

Looking now. Thanks for the ping!

ntsekouras commented:
The sanitization [fixes have been merged](https://github.com/WordPress/gutenberg/pull/39970) in GB, so we should update this PR with the changes from that one.

hellofromtonya commented:
>The sanitization https://github.com/WordPress/gutenberg/pull/39970 in GB, so we should update this PR with the changes from that one.

~Thanks! I'll grab those changes and apply them to this PR .. shortly.~ Backported latest changes from https://github.com/WordPress/gutenberg/pull/39970 in https://github.com/WordPress/wordpress-develop/pull/2488/commits/e51fd0c51ea297bbc717cb42b344b956d0f3f8b4.

hellofromtonya committed:
Backporting latest changes from https://github.com/WordPress/gutenberg/pull/39970.

ntsekouras reviewed (approved):
Regarding the custom taxonomies changes, this looks good to me!

peterwilsoncc commented:
Code changes look good pending fixes to the tests. 

Let's keep the existing blocks in the test suite (ie with tagId, etc) and modify the expected result. The new format can go in as additional tests.

(I say this in the full knowledge it might be part of your own plan)

hellofromtonya committed:
Backport lib/compat/wordpress-6.0/block-patterns.php.

hellofromtonya committed:
Backports /lib/compat/wordpress-6.0/class-gutenberg-rest-pattern-directory-controller.php.

hellofromtonya committed:
Backport lib/compat/wordpress-6.0/class-wp-rest-block-patterns-controller.php.

hellofromtonya committed:
/lib/compat/wordpress-6.0/class-wp-rest-block-pattern-categories-controller.php

hellofromtonya commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-pattern-categories-controller.php:
```diff
@@ -0,0 +1,150 @@
+	 * @since 6.0.0
+	 */
+	public function __construct() {
+		$this->namespace = '__experimental';
```
@TimothyBJacobs @spacedmonkey what should this namespace be? `wp/v2`?

gziolo replied:
It has to. @spacedmonkey or @TimothyBJacobs would know best.

hellofromtonya replied:
Updated in https://github.com/WordPress/wordpress-develop/pull/2488/commits/ceedc0a33b0b336e9465bf6322f4464c28f622c8.

hellofromtonya commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-patterns-controller.php:
```diff
@@ -0,0 +1,198 @@
+	 * @since 6.0.0
+	 */
+	public function __construct() {
+		$this->namespace = '__experimental';
```
@TimothyBJacobs @spacedmonkey what should this namespace be? `wp/v2`?

hellofromtonya replied:
Updated in https://github.com/WordPress/wordpress-develop/pull/2488/commits/ceedc0a33b0b336e9465bf6322f4464c28f622c8.

hellofromtonya committed:
Backports new test classes.

hellofromtonya commented:
This PR captures block patterns and adding custom taxonomies filtering as changes are intermingled in the files. See the list of files backported in the description.

@ntsekouras @ironprogrammer Can you double-check that all changes from the files listed are included in this PR please? Some of the changes "seemed" specific to the Gutenberg plugin; so I didn't include them.

>Let's keep the existing blocks in the test suite (ie with tagId, etc) and modify the expected result. The new format can go in as additional tests.

Tests are failing. I'll fix those once confirmed that all changes to be backported from those files are here in this PR.

ironprogrammer commented:
In response to @hellofromtonya:

> ...Some of the changes "seemed" specific to the Gutenberg plugin; so I didn't include them.

I concur that `lib/compat/wordpress-6.0/block-patterns-update.php` applies to Gutenberg only.

> Tests are failing. I'll fix those once confirmed that all changes to be backported from those files are here in this PR.

Tests needed the new routes to be registered. Please see [PR 3](https://github.com/hellofromtonya/wordpress-develop/pull/3) above, which includes updates that get those working. Please review the amendments and feel free to slurp into your PR any changes you agree with.

Please note that for lack of spelunking time today, I did not address the failure for `Tests_REST_WpRestBlockPatternsController::test_get_items`. It appears the class-level fixture used to clear the Patterns Registry is getting appended to with additional patterns prior to the unit test, which is causing the failure.

ironprogrammer commented on src/wp-includes/block-patterns.php:
```diff
@@ -127,3 +127,205 @@ function _load_remote_featured_patterns() {
+	 * Register patterns for the active theme. If the theme is a child theme,
+	 * let it override any patterns from the parent theme that shares the same slug.
+	 */
+	$themes     = array();
```
👍🏻

ironprogrammer commented on src/wp-includes/block-patterns.php:
```diff
@@ -127,3 +127,205 @@ function _load_remote_featured_patterns() {
+	 * Register patterns for the active theme. If the theme is a child theme,
+	 * let it override any patterns from the parent theme that shares the same slug.
+	 */
+	$themes     = array();
+	$stylesheet = get_stylesheet();
+	$template   = get_template();
+	if ( $stylesheet !== $template ) {
+		$themes[] = wp_get_theme( $stylesheet );
+	}
+	$themes[] = wp_get_theme( $template );
+
+	foreach ( $themes as $theme ) {
```
👍🏻

adamziel commented on src/wp-includes/rest-api/endpoints/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -80,6 +80,7 @@ public function get_items_permissions_check( $request ) {
 	 * Search and retrieve block patterns metadata
 	 *
 	 * @since 5.8.0
+	 * @since 6.0.0 Added 'slug' to request.
```
In the Gutenberg plugin, this class extends `WP_REST_Pattern_Directory_Controller`:

https://github.com/WordPress/gutenberg/blob/cab8cc52cf715722d8d8aa081ee66744e991ab27/lib/compat/wordpress-6.0/class-gutenberg-rest-pattern-directory-controller.php#L12

In the backport, however, it extends `WP_REST_Controller`. Is this intentional? Do they both work in the same way?

adamziel replied:
Ah, I see that the Gutenberg version extends this exact class, presumably to override some functionality which can be just merged right into the code in wordpress-develop.

hellofromtonya replied:
Good question @adamziel. 

The Gutenberg plugin extends `WP_REST_Pattern_Directory_Controller` to add functionality to the existing controller that was added in 5.8.0. The inheritance chain in the plugin is: `Gutenberg_REST_Pattern_Directory_Controller` extends `WP_REST_Pattern_Directory_Controller` extends `WP_REST_Controller`. 

This PR merges the changes from `Gutenberg_REST_Pattern_Directory_Controller` into `WP_REST_Pattern_Directory_Controller`.

adamziel commented on tests/phpunit/tests/rest-api/wpRestBlockPatternsController.php:
```diff
@@ -0,0 +1,151 @@
+	 *
+	 * @var string
+	 */
+	const REQUEST_ROUTE = '/__experimental/block-patterns/patterns';
```
This will need to be updated to wp/v2

hellofromtonya replied:
Updated in https://github.com/WordPress/wordpress-develop/pull/2488/commits/ceedc0a33b0b336e9465bf6322f4464c28f622c8.

adamziel commented on tests/phpunit/tests/rest-api/wpRestBlockPatternCategoriesController.php:
```diff
@@ -0,0 +1,134 @@
+	 *
+	 * @var string
+	 */
+	const REQUEST_ROUTE = '/__experimental/block-patterns/categories';
```
The namespace will need to be updated to wp/v2

hellofromtonya replied:
Updated in https://github.com/WordPress/wordpress-develop/pull/2488/commits/ceedc0a33b0b336e9465bf6322f4464c28f622c8.

adamziel commented:
I left some notes, but other than what's already been said – it looks pretty solid. Thank you @hellofromtonya !

ironprogrammer committed:
Load files, use stable namespace, register routes, girl scouting.

* Update to stable namespace.
* Register routes for WP_REST_Block_Pattern_Categories_Controller
* Require class file for WP_REST_Block_Pattern_Categories_Controller
* Update expected route for WP_REST_Block_Patterns_Controller
* Require class file for WP_REST_Block_Patterns_Controller
* Register routes for WP_REST_Block_Patterns_Controller
* Girl scouting annotations and formatting

hellofromtonya committed:
Fix instance property reflection error in tests.

Resolves the `ReflectionException: Class [name of test class] does not have a property named instance`.

Error occurred as `getProperty()` is not available on `private` or `protected` properties.

This commit uses `ReflectionProperty` to change the `instance` property to `public` for the test.
The property is restored to `private` when done.

hellofromtonya committed:
Fix phpcs whitespace at end of line.

hellofromtonya committed:
Fixes expected query results of Tests_Blocks_wpBlock::test_build_query_vars_from_query_block().

hellofromtonya committed:
Fix arrow alignment in test for phpcs.

spacedmonkey commented:
As this contains REST API, someone ( including me ) for the REST API maintainer team should really review this before it get merged into core.

hellofromtonya committed:
Adds new endpoints to schema setup test.

hellofromtonya committed:
Backport new WP_Theme_JSON::get_patterns() method.

hellofromtonya committed:
Fixes Tests_REST_WpRestBlockPatternsController::test_get_items() test.

TimothyBJacobs commented on src/wp-includes/block-patterns.php:
```diff
@@ -127,3 +127,205 @@ function _load_remote_featured_patterns() {
+		return;
+	}
+
+	if ( ! apply_filters( 'should_load_remote_block_patterns', true ) ) {
```
This needs a doc-block

ironprogrammer replied:
✅ See f5c57d08b1.

TimothyBJacobs commented on src/wp-includes/block-patterns.php:
```diff
@@ -127,3 +127,205 @@ function _load_remote_featured_patterns() {
+	}
+
+	$request         = new WP_REST_Request( 'GET', '/wp/v2/pattern-directory/patterns' );
+	$request['slug'] = implode( ',', $pattern_settings );
```
The `implode` is unnecessary. The API can accept real arrays.

anton-vlasenko replied:
Hm, I get this error if I replace

anton-vlasenko replied:
I get the following error if I replace
```php
$request['slug'] = implode( ',', $pattern_settings );
```
with
```php
$request['slug'] = $pattern_settings;
```
:
```
Uncaught TypeError: rawurlencode(): Argument #1 ($string) must be of type string, array given in /src/wp-content/plugins/gutenberg/lib/compat/wordpress-6.0/class-gutenberg-rest-pattern-directory-controller.php:79
```
This error doesn't occur if `implode` gets called.

anton-vlasenko replied:
I think you are right. `Gutenberg_REST_Pattern_Directory_Controller` has to be fixed to accept arrays instead of using `implode` to concatenate values.
I will submit a PR shortly.
Please ignore my previous message.

anton-vlasenko replied:
I've submitted a PR that fixes this issue.
I would appreciate your review, @TimothyBJacobs.
cc: @spacedmonkey
Trac issue: https://core.trac.wordpress.org/ticket/55574
Github PR: https://github.com/WordPress/wordpress-develop/pull/2591

TimothyBJacobs commented on src/wp-includes/block-patterns.php:
```diff
@@ -127,3 +127,205 @@ function _load_remote_featured_patterns() {
+					$pattern_data = get_file_data( $file, $default_headers );
+
+					if ( empty( $pattern_data['slug'] ) ) {
+						trigger_error(
```
I think these should all be `_doing_it_wrong`s.

ironprogrammer replied:
✅ See faa6dd21ba.

TimothyBJacobs commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-pattern-categories-controller.php:
```diff
@@ -0,0 +1,150 @@
+					'description' => __( 'The category name.' ),
+					'type'        => 'string',
+					'readonly'    => true,
+					'context'     => array( 'view', 'embed' ),
```
For consistency, we should include `edit` here even if we don't currently have a differening response for the `edit` context. Otherwise a user will get no data back if they request the resource in the `edit` context.

ironprogrammer replied:
✅ See c5994cda22.

TimothyBJacobs commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-pattern-categories-controller.php:
```diff
@@ -0,0 +1,150 @@
+		$data    = $this->add_additional_fields_to_object( $data, $request );
+		$data    = $this->filter_response_by_context( $data, $context );
+
+		return rest_ensure_response( $data );
```
This should typically be filterable. It looks like the Pattern Registry does allow for more fields than just these two, so it would be usable by developers.

jsnajdr replied:
Which kind of filtering do you have in mind? I can already register an additional field with `register_rest_field( 'block-pattern', ... )`.

TimothyBJacobs commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-patterns-controller.php:
```diff
@@ -0,0 +1,198 @@
+	 */
+	public function get_items( $request ) {
+		// Load block patterns from w.org.
+		_load_remote_block_patterns(); // Patterns with the `core` keyword.
+		_load_remote_featured_patterns(); // Patterns in the `featured` category.
+		_register_remote_theme_patterns(); // Patterns requested by current theme.
```
Are these methods guarded against running more than once in a request? Looking at `_load_remote_block_patterns` at least, it seems it will try and register the patterns again. It's very important that the REST API is reentrant.

jsnajdr replied:
They are not guarded. When writing them, I assumed that the handler can ever run only once. But that was wrong. What would be a good way to prevent the `_load_remote_*` being run twice? Some `$already_loaded` global variable?

anton-vlasenko replied:
I've checked it in the XDebug, and these methods don't get called twice during a simple GET `wp/v2/block-patterns/patterns` request.
But I've submitted a PR to fix this possible issue: https://github.com/WordPress/wordpress-develop/pull/2576
cc: @gziolo

gziolo replied:
Excellent, let's wait for confirmation from @TimothyBJacobs on the changes proposed.

jsnajdr replied:
> these methods don't get called twice during a simple GET

During regular requests they probably never are, but what @TimothyBJacobs probably had in mind is that the handler can be called multiple times when doing special things: batching requests, executing the endpoints when preloading, embedding linked requests into one response... In these cases, multiple requests are being processed within one parent one.

anton-vlasenko replied:
Thank you, @jsnajdr.
I understand it.
I was primarily concerned about single requests (fortunately, the issue doesn't affect single requests).
I wouldn't submit a fix if I weren't sure that the issue could affect batch requests etc.

TimothyBJacobs commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-patterns-controller.php:
```diff
@@ -0,0 +1,198 @@
+		$context = ! empty( $request['context'] ) ? $request['context'] : 'view';
+		$data    = $this->add_additional_fields_to_object( $data, $request );
+		$data    = $this->filter_response_by_context( $data, $context );
+		return rest_ensure_response( $data );
```
Same thing about filter support.

TimothyBJacobs commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-patterns-controller.php:
```diff
@@ -0,0 +1,198 @@
+			'name',
+			'title',
+			'description',
+			'viewportWidth',
```
I guess it is is too late now, but the pattern we've been following is for the REST API to use `snake_case`. See the Block Types controller for instance.

gziolo replied:
@ntsekouras, can you check whether it's still possible to update this name of the field?

jsnajdr replied:
I think it should be possible, and implemented it now in this Gutenberg PR: https://github.com/WordPress/gutenberg/pull/40254

The patch changes the REST endpoint to respond with `block_types` and `viewport_width` field, and also changes the `core` data store to convert them back to `blockTypes` and `viewportWidth` when receiving the response. So that consumers of `wp.data.select( 'core' ).getBlockPatterns()` see the same `blockTypes` and `viewportWidth` fields as they did before.

@gziolo What would be the mechanics of including this in WP 6.0 beta? The REST endpoint part can be added as another commit in this PR, but the `wp.data` part needs to get there separately.

gziolo replied:
We are going to cherry-pick https://github.com/WordPress/gutenberg/pull/40254 and https://github.com/WordPress/gutenberg/pull/40252 to `release/13.0` branch in Gutenberg and publish packages to npm in 1-2 hours.

gziolo replied:
Addressed with https://core.trac.wordpress.org/changeset/53154 that was based on https://github.com/WordPress/wordpress-develop/pull/2574.

TimothyBJacobs commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-patterns-controller.php:
```diff
@@ -0,0 +1,198 @@
+					'description' => __( 'The pattern name.' ),
+					'type'        => 'string',
+					'readonly'    => true,
+					'context'     => array( 'view', 'embed' ),
```
As in the previous controller, let's add `edit` to these.

ironprogrammer replied:
✅ See 39ea1e3c64.

ironprogrammer replied:
✅ Also see dcfe4f5553.

TimothyBJacobs commented on tests/phpunit/tests/rest-api/wpRestBlockPatternCategoriesController.php:
```diff
@@ -0,0 +1,145 @@
+			$this->assertSame( $expected_names[ $idx ], $item['name'] );
+			$this->assertSame( $expected_fields, array_keys( $item ) );
+		}
+	}
```
I would like to see a test that verifies the capabilitiy checks are working properly. Typically we include a logged-out user, and a user with subscriber level capabilities.

ironprogrammer replied:
✅ See f7ce9521dc.

TimothyBJacobs commented on tests/phpunit/tests/rest-api/wpRestBlockPatternsController.php:
```diff
@@ -0,0 +1,171 @@
+			$data[1],
+			'WP_REST_Block_Patterns_Controller::get_items() should return test/two'
+		);
+	}
```
Same as the previous controller about a permission test.

ironprogrammer replied:
✅ See 0b89e425e7.

spacedmonkey commented on src/wp-includes/block-patterns.php:
```diff
@@ -127,3 +127,205 @@ function _load_remote_featured_patterns() {
+
+	foreach ( $themes as $theme ) {
+		$dirpath = $theme->get_stylesheet_directory() . '/patterns/';
+		if ( file_exists( $dirpath ) ) {
```
Might be worth doing an is readable check here.

ironprogrammer replied:
✅ See 820ee72738.

ntsekouras replied:
After the [merged changes with the new check](https://github.com/WordPress/wordpress-develop/pull/2567), we need to remove the extra one (`file_exists`), since we do it in the above line.. See this comment: https://github.com/WordPress/wordpress-develop/pull/2567#discussion_r848204070

ironprogrammer commented:
Hi, @TimothyBJacobs and @spacedmonkey -- thank you for your feedback! 🙌🏻 Updates for this last round of comments are being addressed over at https://github.com/WordPress/wordpress-develop/pull/2567 (forked from this PR).

CC @gziolo @adamziel @anton-vlasenko @hellofromtonya

gziolo commented:
I'm still trying to find out how to go about all those backports to WordPress core. https://github.com/WordPress/wordpress-develop/pull/2567 has failing CI jobs. This branch was passing all CI checks so I consider committing it as is and re-apply all other changes as individual commits directly in WP core.

gziolo commented:
I'm getting closer to have https://github.com/WordPress/wordpress-develop/pull/2567 pass all CI checks.

gziolo commented:
I committed all the changes included in https://github.com/WordPress/wordpress-develop/pull/2567 with https://core.trac.wordpress.org/changeset/53152.

We still need to address the reamining feedback listed in this PR.
